File: poker_eval.c
#include "poker_eval.h"

#include <string.h>

static void clear_value(HandValue *value) {
    int i;

    value->category = HAND_HIGH_CARD;
    for (i = 0; i < 5; i++) {
        value->tie_breakers[i] = 0;
    }
}

static int compare_int_desc(const void *left, const void *right) {
    int a = *(const int *)left;
    int b = *(const int *)right;
    return b - a;
}

static void sort_desc(int values[], int count) {
    int i;
    int j;

    for (i = 0; i < count - 1; i++) {
        for (j = i + 1; j < count; j++) {
            if (compare_int_desc(&values[i], &values[j]) > 0) {
                int temp = values[i];
                values[i] = values[j];
                values[j] = temp;
            }
        }
    }
}

static int straight_high_card(const int rank_count[15]) {
    int high;

    for (high = RANK_A; high >= RANK_6; high--) {
        if (rank_count[high] &&
            rank_count[high - 1] &&
            rank_count[high - 2] &&
            rank_count[high - 3] &&
            rank_count[high - 4]) {
            return high;
        }
    }

    if (rank_count[RANK_A] &&
        rank_count[RANK_5] &&
        rank_count[RANK_4] &&
        rank_count[RANK_3] &&
        rank_count[RANK_2]) {
        return RANK_5;
    }

    return 0;
}

static HandValue evaluate_5cards(const Card cards[5]) {
    HandValue value;
    int rank_count[15] = {0};
    int suit_count[4] = {0};
    int ranks[5];
    int kickers[5];
    int pair_ranks[2];
    int straight_high;
    int flush;
    int four = 0;
    int three = 0;
    int pair_count = 0;
    int kicker_count = 0;
    int i;
    int rank;

    clear_value(&value);

    for (i = 0; i < 5; i++) {
        rank_count[cards[i].rank]++;
        suit_count[cards[i].suit]++;
        ranks[i] = cards[i].rank;
    }

    sort_desc(ranks, 5);
    flush = suit_count[cards[0].suit] == 5;
    straight_high = straight_high_card(rank_count);

    if (flush && straight_high) {
        value.category = HAND_STRAIGHT_FLUSH;
        value.tie_breakers[0] = straight_high;
        return value;
    }

    for (rank = RANK_A; rank >= RANK_2; rank--) {
        if (rank_count[rank] == 4) {
            four = rank;
        } else if (rank_count[rank] == 3) {
            three = rank;
        } else if (rank_count[rank] == 2) {
            pair_ranks[pair_count++] = rank;
        } else if (rank_count[rank] == 1) {
            kickers[kicker_count++] = rank;
        }
    }

    if (four) {
        value.category = HAND_FOUR_OF_A_KIND;
        value.tie_breakers[0] = four;
        value.tie_breakers[1] = kickers[0];
        return value;
    }

    if (three && pair_count > 0) {
        value.category = HAND_FULL_HOUSE;
        value.tie_breakers[0] = three;
        value.tie_breakers[1] = pair_ranks[0];
        return value;
    }

    if (flush) {
        value.category = HAND_FLUSH;
        memcpy(value.tie_breakers, ranks, sizeof(ranks));
        return value;
    }

    if (straight_high) {
        value.category = HAND_STRAIGHT;
        value.tie_breakers[0] = straight_high;
        return value;
    }

    if (three) {
        value.category = HAND_THREE_OF_A_KIND;
        value.tie_breakers[0] = three;
        value.tie_breakers[1] = kickers[0];
        value.tie_breakers[2] = kickers[1];
        return value;
    }

    if (pair_count == 2) {
        value.category = HAND_TWO_PAIR;
        value.tie_breakers[0] = pair_ranks[0];
        value.tie_breakers[1] = pair_ranks[1];
        value.tie_breakers[2] = kickers[0];
        return value;
    }

    if (pair_count == 1) {
        value.category = HAND_ONE_PAIR;
        value.tie_breakers[0] = pair_ranks[0];
        value.tie_breakers[1] = kickers[0];
        value.tie_breakers[2] = kickers[1];
        value.tie_breakers[3] = kickers[2];
        return value;
    }

    value.category = HAND_HIGH_CARD;
    memcpy(value.tie_breakers, ranks, sizeof(ranks));
    return value;
}

HandValue evaluate_7cards(const Card cards[7]) {
    HandValue best;
    int initialized = 0;
    int a;
    int b;
    int c;
    int d;
    int e;

    clear_value(&best);

    for (a = 0; a < 3; a++) {
        for (b = a + 1; b < 4; b++) {
            for (c = b + 1; c < 5; c++) {
                for (d = c + 1; d < 6; d++) {
                    for (e = d + 1; e < 7; e++) {
                        Card combo[5] = {
                            cards[a], cards[b], cards[c], cards[d], cards[e]
                        };
                        HandValue current = evaluate_5cards(combo);

                        if (!initialized || compare_hand_values(current, best) > 0) {
                            best = current;
                            initialized = 1;
                        }
                    }
                }
            }
        }
    }

    return best;
}

int compare_hand_values(HandValue a, HandValue b) {
    int i;

    if (a.category > b.category) {
        return 1;
    }
    if (a.category < b.category) {
        return -1;
    }

    for (i = 0; i < 5; i++) {
        if (a.tie_breakers[i] > b.tie_breakers[i]) {
            return 1;
        }
        if (a.tie_breakers[i] < b.tie_breakers[i]) {
            return -1;
        }
    }

    return 0;
}

const char *hand_category_to_string(HandCategory category) {
    switch (category) {
        case HAND_STRAIGHT_FLUSH: return "Straight flush";
        case HAND_FOUR_OF_A_KIND: return "Four of a kind";
        case HAND_FULL_HOUSE: return "Full house";
        case HAND_FLUSH: return "Flush";
        case HAND_STRAIGHT: return "Straight";
        case HAND_THREE_OF_A_KIND: return "Three of a kind";
        case HAND_TWO_PAIR: return "Two pair";
        case HAND_ONE_PAIR: return "One pair";
        case HAND_HIGH_CARD: return "High card";
        default: return "Unknown";
    }
}

File: poker_eval.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum

from common import Card, Rank


class HandCategory(IntEnum):
    HIGH_CARD = 0
    ONE_PAIR = 1
    TWO_PAIR = 2
    THREE_OF_A_KIND = 3
    STRAIGHT = 4
    FLUSH = 5
    FULL_HOUSE = 6
    FOUR_OF_A_KIND = 7
    STRAIGHT_FLUSH = 8


@dataclass(frozen=True)
class HandValue:
    category: HandCategory
    tie_breakers: tuple[int, ...] = field(default_factory=tuple)


def _straight_high_card(rank_counts: list[int]) -> int:
    for high in range(int(Rank.ACE), int(Rank.SIX) - 1, -1):
        if (
            rank_counts[high] > 0
            and rank_counts[high - 1] > 0
            and rank_counts[high - 2] > 0
            and rank_counts[high - 3] > 0
            and rank_counts[high - 4] > 0
        ):
            return high

    if (
        rank_counts[int(Rank.ACE)] > 0
        and rank_counts[int(Rank.FIVE)] > 0
        and rank_counts[int(Rank.FOUR)] > 0
        and rank_counts[int(Rank.THREE)] > 0
        and rank_counts[int(Rank.TWO)] > 0
    ):
        return int(Rank.FIVE)

    return 0


def _top_ranks(rank_counts: list[int], count: int, exclude: tuple[int, ...] = ()) -> tuple[int, ...]:
    ranks: list[int] = []
    excluded = set(exclude)

    for rank in range(int(Rank.ACE), int(Rank.TWO) - 1, -1):
        if rank in excluded:
            continue
        if rank_counts[rank] > 0:
            ranks.append(rank)
            if len(ranks) == count:
                break

    return tuple(ranks)


def evaluate_7cards(cards: list[Card] | tuple[Card, ...]) -> HandValue:
    if len(cards) != 7:
        raise ValueError("evaluate_7cards expects exactly 7 cards")

    rank_counts = [0] * 15
    suited_ranks: list[list[int]] = [[], [], [], []]

    for card in cards:
        rank = int(card.rank)
        suit = int(card.suit)
        rank_counts[rank] += 1
        suited_ranks[suit].append(rank)

    # Straight flush
    for ranks in suited_ranks:
        if len(ranks) >= 5:
            suited_counts = [0] * 15
            for rank in ranks:
                suited_counts[rank] += 1
            straight_flush_high = _straight_high_card(suited_counts)
            if straight_flush_high:
                return HandValue(HandCategory.STRAIGHT_FLUSH, (straight_flush_high,))

    quads: list[int] = []
    trips: list[int] = []
    pairs: list[int] = []

    for rank in range(int(Rank.ACE), int(Rank.TWO) - 1, -1):
        count = rank_counts[rank]
        if count == 4:
            quads.append(rank)
        elif count == 3:
            trips.append(rank)
        elif count == 2:
            pairs.append(rank)

    # Four of a kind
    if quads:
        quad = quads[0]
        kicker = _top_ranks(rank_counts, 1, (quad,))[0]
        return HandValue(HandCategory.FOUR_OF_A_KIND, (quad, kicker))

    # Full house. If there are two trips, the lower trip can be used as the pair.
    if trips and (len(trips) >= 2 or pairs):
        trip = trips[0]
        pair = trips[1] if len(trips) >= 2 else pairs[0]
        return HandValue(HandCategory.FULL_HOUSE, (trip, pair))

    # Flush
    best_flush: tuple[int, ...] | None = None
    for ranks in suited_ranks:
        if len(ranks) >= 5:
            flush_ranks = tuple(sorted(ranks, reverse=True)[:5])
            if best_flush is None or flush_ranks > best_flush:
                best_flush = flush_ranks
    if best_flush is not None:
        return HandValue(HandCategory.FLUSH, best_flush)

    # Straight
    straight_high = _straight_high_card(rank_counts)
    if straight_high:
        return HandValue(HandCategory.STRAIGHT, (straight_high,))

    # Three of a kind
    if trips:
        trip = trips[0]
        kickers = _top_ranks(rank_counts, 2, (trip,))
        return HandValue(HandCategory.THREE_OF_A_KIND, (trip, *kickers))

    # Two pair
    if len(pairs) == 2:
        kicker = _top_ranks(rank_counts, 1, (pairs[0], pairs[1]))[0]
        return HandValue(HandCategory.TWO_PAIR, (pairs[0], pairs[1], kicker))

    if len(pairs) >= 3:
        kicker = _top_ranks(rank_counts, 1, (pairs[0], pairs[1]))[0]
        return HandValue(HandCategory.TWO_PAIR, (pairs[0], pairs[1], kicker))

    # One pair
    if len(pairs) == 1:
        pair = pairs[0]
        kickers = _top_ranks(rank_counts, 3, (pair,))
        return HandValue(HandCategory.ONE_PAIR, (pair, *kickers))

    return HandValue(HandCategory.HIGH_CARD, _top_ranks(rank_counts, 5))


def compare_hand_values(a: HandValue, b: HandValue) -> int:
    a_key = _hand_value_key(a)
    b_key = _hand_value_key(b)

    if a_key > b_key:
        return 1
    if a_key < b_key:
        return -1
    return 0


def _hand_value_key(value: HandValue) -> tuple[int, tuple[int, ...]]:
    return int(value.category), value.tie_breakers


def hand_category_to_string(category: HandCategory) -> str:
    names = {
        HandCategory.STRAIGHT_FLUSH: "Straight flush",
        HandCategory.FOUR_OF_A_KIND: "Four of a kind",
        HandCategory.FULL_HOUSE: "Full house",
        HandCategory.FLUSH: "Flush",
        HandCategory.STRAIGHT: "Straight",
        HandCategory.THREE_OF_A_KIND: "Three of a kind",
        HandCategory.TWO_PAIR: "Two pair",
        HandCategory.ONE_PAIR: "One pair",
        HandCategory.HIGH_CARD: "High card",
    }
    return names.get(category, "Unknown")
